packages/@react-stately/table/src/useTableColumnResizeState.ts

```diff
@@ -41,33 +41,39 @@ export interface TableColumnResizeStateProps {
   /** Callback that is invoked when the resize event is ended. */
   onColumnResizeEnd?: (affectedColumnWidths: AffectedColumnWidths) => void,
   /** The default table width. */
-  tableWidth?: number
+  tableWidth?: number,
+  onResize?: (inRender: boolean) => void
 }
 
 interface ColumnState<T> {
   columns: GridNode<T>[]
 }
 
 export function useTableColumnResizeState<T>(props: TableColumnResizeStateProps, state: ColumnState<T>): TableColumnResizeState<T> {
-  const {getDefaultWidth, tableWidth: defaultTableWidth = null} = props;
+  const {getDefaultWidth, tableWidth: defaultTableWidth = null, onResize} = props;
   const {columns} = state;
   const columnsRef = useRef<GridNode<T>[]>([]);
   const tableWidth = useRef<number>(defaultTableWidth);
   const isResizing = useRef<boolean>(null);
   const startResizeContentWidth = useRef<number>();
 
-  const [columnWidths, setColumnWidths] = useState<Map<Key, number>>(new Map(columns.map(col => [col.key, 0])));
-  const columnWidthsRef = useRef<Map<Key, number>>(columnWidths);
+  const columnWidthsRef = useRef<Map<Key, number>>(new Map(columns.map(col => [col.key, 0])));
+  const {1: setColumnWidths} = useState<Map<Key, number>>(columnWidthsRef.current);
   const affectedColumnWidthsRef = useRef<AffectedColumnWidths>([]);
   const [resizedColumns, setResizedColumns] = useState<Set<Key>>(new Set());
   const resizedColumnsRef = useRef<Set<Key>>(resizedColumns);
 
   const [currentlyResizingColumn, setCurrentlyResizingColumn] = useState<Key>(null);
 
-  function setColumnWidthsForRef(newWidths: Map<Key, number>) {
+  let prevWidths = useRef<Map<Key, number>>(columnWidthsRef.current);
+  function setColumnWidthsForRef(newWidths: Map<Key, number>, inRender = false) {
     columnWidthsRef.current = newWidths;
-    // new map so that change detection is triggered
-    setColumnWidths(newWidths);
+    if (prevWidths.current !== newWidths && onResize) {
```

Comment:
Non issue in functionality, just something I was thinking about:
Looking at this again, is there a scenario where `newWidths` might be the same Map object as the one stored in `prevWidths`? Seems like we create new Map objects every time `setColumnWidthsForRef` is called. If this is more  of a guard against the possibility of `newWidths` being the same Map as `prevWidths` then shouldn't we set `prevWidths.current` outside the if at the end of this function?

Reply:
prevWidths only matters in the case where you manage the state update outside the hook, it's just to a defensive coding for calling onResize multiple times with the same value since we don't know how the state updates would be handled outside. When it's inside, setState already performs this check and will only cause a render for new Maps.